proposals/dark_mode_extension_icons.md

```diff
@@ -0,0 +1,351 @@
+1. Improved icon visibility on management the management and shortcuts pages.
+1. Dark mode icon declarations made possible through the extension manifest.
+1. setIcon() will allow setting of dark and/or light mode specific icons.
+1. Improved icon visibility on context menu (Chrome and Safari reply on default
+icons, only Firefox can specify icons).
+1. Improved icon visibility on side panel (Chrome replies on default icons,
+Firefox can specify icons).
+
+### Known Consumers
+
+The Chromium bug has a significant amount of developer interest.
+
+## Specification
+
+### Schema
+
+// TODO: Obtain consensus on any of the following options before merging.
+
+Incumbent action.setIcon(), for reference.
+```
+action.setIcon({
+  path,
+  tabId
+  imageData,
+});
+```
+
+---
+
+#### Option (donut option)
+
+Don't implement setIcon() for dark mode, as it's used for dynamic icons anyway.
+
+* Pros: Easy to implement.
+* Cons: Developers would need to find their own ways of detecting dark mode.
+
+---
+
+#### Option (combined option)
+
+manifest.json
+```
+{
+  "icon_variants": {
+    "dark": {
+      "128": "128_dark.png"
+    },
+    "light": {
+      "128": "128_light.png"
+    }
+  },
+  "action": {
+    "default_icon": {
+      "128": "128_action.png"
+    },
+    "icon_variants": {
+      "dark": {
+        "128": "128_action_dark.png"
+      },
+      "light": {
+        "128": "128_action_light.png"
+      }
+    }
+  }
+}
+```
+
+action.setIcon()
+```
+action.setIcon({
+  variants: {
+    "dark": {
+      "128": "128_action_dark.png"
+    },
+    "light": {
+      "128": LightImageData
+    }
+  },
+});
+```
+
+Schema
+
+manifest.json
+```
+// `Mode` is dark or light (if either is used, both are expected).
+const Modes: string[] = ["dark", "light];
+const Mode = Modes[Math.floor(Math.random() * Modes.length())];
+
+// Primitive type declaration.
+type Size = string;
+type Path = string;
+
+// Map icon size to path.
+const IconSizeToPath: map<Size, Path>;
+
+// Map mode to a string path or a dictionary mapping icon size to a string path.
+const IconVariants: map<Mode, IconSizeToPath | Path>;
+
+{
+  ...
+  "icon_variants": IconVariants;
+  "action": {
+    ...
+    "icon_variants": IconVariants
+  },
+};
+```
+
+action.setIcon()
+```
+// ImageData is an interface representing canvas element pixels.
+const IconSizeToImageData: map<Size, Path>;
+const IconVariantsWithImageData:
+  map<Mode, IconSizeToPath | IconSizeToImageData | ImageData | Path>;
+const ImageDataDictionary: map<Size, ImageData>;
+
+action.setIcon(
+  ...
+  icon_variants?: IconVariantsWithImageData,
+);
+```
+
+* Pros: Easy for developers to use.
+* Cons: Complex to define and implement.
+
+
+---
+
+#### Option (flat option)
+
+manifest.json
+```
+"icon_variants": [{
+  "16": "16.png",
+  "32": "32.png",
+}, {
+  "16": "dark16.png",
+  "32": "dark32.png",
+  "color_scheme": "dark",
+}, {
+  "16": "light16.png",
+  "32": "light32.png",
+  "color_scheme": "light"
+}]
+```
```

Comment:
I think this option should allow `light` to be assumed for the case that is missing `color_scheme`.

```json
[{
  "16": "16.png",
  "32": "32.png",
}, {
  "16": "dark16.png",
  "32": "dark32.png",
  "color_scheme": "dark",
}]
```

Also this option should allow `"color_scheme": ["light", "dark"]` like the CSS property, so a universal icon that works in both modes can be set once. This lets the developer use the new property / methods, and not need to fallback to the old `icons` key.

```json
[{
  "16": "16.png",
  "32": "32.png",
  "color_scheme": ["light", "dark"]
}]
```

Reply:
Only other note, we should allow an `"any"` size in this new proposal, so SVG is a first class citizen. At least carve it out so `"any"` is not an error and can be supported by browsers that support SVG icons. (This matches what Web App Manifest does.)

Reply:
In this flatten option, if we follow [PWA manifest format](https://developer.mozilla.org/en-US/docs/Web/Manifest/icons), it should use `sizes` as the key, rather than "16", "32", "any" as keys.
```
[
  {
    "src": "icon/48.png",
    "sizes": "48x48"
  },
  {
    "src": "icon/icon.svg",
    "sizes": "any"
  }
]
```

I like your "color_scheme" suggestion.

Reply:
I don't think we need to completely follow the Web App Manifest format here. I just wanted `any` to be considered. The web app format is too verbose in my opinion, and does a poor job of grouping similar icons. Their syntax is mostly severed by the need to support varying image formats, where our need is mostly about varying color schemes. We also have the benefit of using file extensions for type info, where they needed a `type` key.

Reply:
@xeenon, is there a reason you would like to support only defining a light icon? If not, could we make the default `"color_scheme": ["light", "dark"]` instead of just light? That would allow you to omit the key in the case where the icon works for both and we wouldn't need to support an array there.

Reply:
`any` has been included. So has ["dark", "light"].

Reply:
@hanguokai In the past during one of the meetings I proposed to use the same syntax as the WebManifest. This was considered too verbose. Which I have to agree with. Seems the syntax used by the WebManifest is designed to match `<link rel="icon">`. This is why I proposed the hybrid between what we had in webExtensions and webManifests. As many of the descriptors (`color_scheme`, `type`)  applies to a group of icons.

Reply:
@xeenon Have to agree with @oliverdunk on making `"color_scheme": ["light", "dark"]` the default. Historically extension icons have suggested by at least Google Chrome to be designed to work for both `light` and `dark` extension themes. @xeenon What is the use-case for providing an array instead of simply omitting it? For additional code clarity or do you expect additional color_schemes to be added in the future?

Reply:
@carlosjeurissen Yeah, array for future proofing really, and it matches the `color-scheme` CSS property supporting multiple values.

Reply:
@xeenon Say additional items will be added to "color_scheme" in the future (which I do not expect), how would the browser handle this? Should it simply ignore schemes it does not understand?

Still not convinced why having `color_scheme` set to `["light", "dark"]` has any value. I feel like omitting it would say "do not filter on color_scheme".

Reply:
@xeenon Convinced. Thanks for the detailed explanation. All in favor for future proofing ourselves. We should attempt to not require manifest version upgrades as much as possible.